### ISSUE TITLE
[Go] Clean up ExternalOp comparisons

### ISSUE THREAD

**Issue by gj** [osohq/oso#795]
When handling an `ExternalOp` event, the Go adapter currently relies on operands implementing the `Comparer` interface defined here:

https://github.com/osohq/oso/blob/992b72881dfe528a116630573941eaa816e5900a/languages/go/interfaces/interfaces.go#L6-L11

If one of the operands doesn't implement `Comparer` — for example, if it's a type that doesn't have an `Equal()` method, such as a new type arising from a type alias — then an error is thrown.

As a quick fix, we should probably add the same `reflect.DeepEqual()` fallback that we have in the unification code path…

https://github.com/osohq/oso/blob/992b72881dfe528a116630573941eaa816e5900a/languages/go/internal/host/host.go#L192-L197

…to the equality checks we perform in the ExternalOp code path:

https://github.com/osohq/oso/blob/992b72881dfe528a116630573941eaa816e5900a/languages/go/query.go#L248-L263

However that's just for the `Equal()` piece of the `Comparer` interface — I'm not sure what to do for a generic `Lt()` fallback, or if one even makes sense to begin with.

### Alternatives

- Catch the runtime error and throw our own (more understandable/instructive) error telling users to implement `Equal()` / `Lt()` for whatever type is missing the method(s)
- Be explicit in the (forthcoming) Go docs that users must ensure their application types implement the `Comparer` interface.

**Comment by samscott89** [comment]
Adding on to this, the Comparer interface itself is a bit rough:

```go
type Foo string

func (left Foo) Equal(right Foo) bool {
    // ...
}
```

isn't sufficient to compare `Foo`s with a string because: 

- The input type needs to match exactly (i.e. be other Comparer)
- We don't currently support comparing instances with not instances. So if you do foo = "abc" where foo is an instance of `Foo`, it's not going to work

First it needs to be:

```go
type Foo string
func (left Foo) Equal(right Comparer) bool {
    if other, ok := right.(string); ok    {
		return left == other
	}
}
```

Secondly we need to fix (I think this is true across all languages) comparing native + host types.

**Comment by gj** [comment]
@gw000 or @samscott89 — was this fixed by #1023?

**Comment by samscott89** [comment]
Yes!
